3scaleAdapter/pkg/threescale/threescale.go

```diff
@@ -50,6 +49,110 @@ type (
 // For this PoC I'm using the authorize template, but we should check if the quota template
 // is more convenient and we can do some optimizations around.
 var _ authorization.HandleAuthorizationServiceServer = &Threescale{}
+var _ logentry.HandleLogEntryServiceServer = &Threescale{}
+
+func (s *Threescale) HandleLogEntry(ctx context.Context, l *logentry.HandleLogEntryRequest) (*v1beta1.ReportResult, error) {
+	var report v1beta1.ReportResult
+	var err error
+
+	if l.AdapterConfig == nil {
+		return &report, fmt.Errorf("adapter config not available")
+	}
+
+	cfg := &config.Params{}
+	if err := cfg.Unmarshal(l.AdapterConfig.Value); err != nil {
+		log.Errorf("error unmarshalling adapter config: %v", err)
+		return &report, err
+	}
+
+	systemC, err := s.systemClientBuilder(cfg.SystemUrl)
+
+	if err != nil {
+		log.Errorf("error building HTTP client for 3scale system - %s", err.Error())
+		return &report, err
+	}
+
+	err = s.ReportUsage(cfg, l.Instances, systemC)
+
+	if err != nil {
+		log.Errorf("error reporting usage - %s", err.Error())
+	}
+
+	return &report, err
+}
+func (s *Threescale) ReportUsage(cfg *config.Params, instances []*logentry.InstanceMsg, c *sysC.ThreeScaleClient) (error) {
+
+	for _, instance := range instances {
+
+		urlWithPath := instance.Variables["url"].GetStringValue()
+		userKey := instance.Variables["user"].GetStringValue()
+		method := instance.Variables["method"].GetStringValue()
+
+		//TODO: Report proper timestamp
+		//timestamp := request.Timestamp.GetValue()
+
+		u, err := url.Parse(urlWithPath)
+		if err != nil {
+			return err
+		}
+
+		if urlWithPath == "" || userKey == "" || method == "" {
+			return errors.New("missing required parameters")
+		}
+
+		// TODO - Some caching around this
+		pce, err := c.GetLatestProxyConfig(cfg.AccessToken, cfg.ServiceId, "production")
+		if err != nil {
+			return err
+		}
+
+		authType := pce.ProxyConfig.Content.BackendAuthenticationType
+
+		switch authType {
+		case "provider_key", "service_token":
+			return s.doReport(cfg.ServiceId, userKey, u.Path, method, pce.ProxyConfig)
+		default:
+			return fmt.Errorf("unsupported auth type for service %s", cfg.ServiceId)
+
+		}
+	}
+	return nil
+}
+
+func (s *Threescale) doReport(svcID string, userKey string, path string, method string, conf sysC.ProxyConfig) (error) {
+
```

Comment:
I think this function should return `rpc.Status` as before and propagate errors back to the user, wdyt?